File: health_sync/providers/strava.py
from __future__ import annotations

import json
import secrets
from datetime import UTC, datetime
from urllib.parse import quote, urlencode, urlparse

import requests

from ..config import LoadedConfig, require_str
from ..db import HealthSyncDb
from ..util import (
    dt_to_iso_z,
    iso_to_dt,
    oauth_listen_for_code,
    open_in_browser,
    parse_yyyy_mm_dd,
    request_json,
    sha256_hex,
    to_epoch_seconds,
    utc_now_iso,
)

STRAVA_PROVIDER = "strava"
STRAVA_BASE = "https://www.strava.com/api/v3"
STRAVA_OAUTH_AUTHORIZE = "https://www.strava.com/oauth/authorize"
STRAVA_OAUTH_TOKEN = "https://www.strava.com/oauth/token"


def _strava_default_redirect_uri() -> str:
    return "http://127.0.0.1:8486/callback"


def _strava_redirect(cfg: LoadedConfig) -> tuple[str, int, str]:
    redirect_uri = cfg.config.strava.redirect_uri or _strava_default_redirect_uri()
    u = urlparse(redirect_uri)
    if u.scheme not in ("http", "https"):
        raise RuntimeError(f"Invalid `strava.redirect_uri`: {redirect_uri}")
    return redirect_uri, u.port or (443 if u.scheme == "https" else 80), u.path or "/callback"


def _strava_scopes(cfg: LoadedConfig) -> str:
    raw = cfg.config.strava.scopes
    seen: set[str] = set()
    out = [p for p in (x.strip() for x in raw.replace(" ", ",").split(",")) if p and (p not in seen and not seen.add(p))]
    return ",".join(out) if out else raw


def _strava_expires_to_iso(v: object) -> str | None:
    try:
        return dt_to_iso_z(datetime.fromtimestamp(int(v), tz=UTC)) if v is not None else None
    except Exception:  # noqa: BLE001
        return None


def _strava_expires_to_dt(v: str | None) -> datetime | None:
    if not v:
        return None
    s = v.strip()
    if not s:
        return None
    try:
        return datetime.fromtimestamp(int(s), tz=UTC) if s.isdigit() else iso_to_dt(s)
    except Exception:  # noqa: BLE001
        return None


def _store_token(db: HealthSyncDb, token: dict[str, object], refresh_fallback: str | None = None) -> str:
    access = str(token["access_token"])
    refresh = str(token.get("refresh_token") or refresh_fallback or "") or None
    db.set_oauth_token(
        provider=STRAVA_PROVIDER,
        access_token=access,
        refresh_token=refresh,
        token_type=str(token.get("token_type") or "Bearer"),
        scope=(str(token["scope"]) if token.get("scope") is not None else None),
        expires_at=_strava_expires_to_iso(token.get("expires_at")),
        extra={
            k: v
            for k, v in token.items()
            if k not in {"access_token", "refresh_token", "token_type", "scope", "expires_at"}
        },
    )
    return access


def strava_auth(db: HealthSyncDb, cfg: LoadedConfig, *, listen_host: str = "127.0.0.1", listen_port: int = 0) -> None:
    if cfg.config.strava.access_token:
        db.set_oauth_token(
            provider=STRAVA_PROVIDER,
            access_token=cfg.config.strava.access_token,
            refresh_token=None,
            token_type="Bearer",
            scope=None,
            expires_at=None,
            extra={"method": "static_access_token"},
        )
        print("Stored Strava access token in DB.")
        return

    client_id = require_str(cfg, cfg.config.strava.client_id, key="strava.client_id")
    client_secret = require_str(cfg, cfg.config.strava.client_secret, key="strava.client_secret")
    redirect_uri, redirect_port, callback_path = _strava_redirect(cfg)
    listen_port = listen_port or redirect_port

    state = secrets.token_urlsafe(16)
    auth_url = f"{STRAVA_OAUTH_AUTHORIZE}?{urlencode({'client_id': client_id, 'response_type': 'code', 'redirect_uri': redirect_uri, 'approval_prompt': cfg.config.strava.approval_prompt or 'auto', 'scope': _strava_scopes(cfg), 'state': state}, quote_via=quote)}"
    print("Open this URL to authorize Strava:")
    print(auth_url)
    open_in_browser(auth_url)

    res = oauth_listen_for_code(listen_host=listen_host, listen_port=listen_port, callback_path=callback_path)
    if res.error:
        raise RuntimeError(f"Strava auth error: {res.error}")
    if res.state and res.state != state:
        raise RuntimeError("Strava auth failed: state mismatch")

    token = request_json(
        requests.Session(),
        "POST",
        STRAVA_OAUTH_TOKEN,
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        data={
            "client_id": client_id,
            "client_secret": client_secret,
            "code": res.code,
            "grant_type": "authorization_code",
        },
    )
    _store_token(db, token)
    print("Stored Strava OAuth token in DB.")


def _strava_refresh_if_needed(db: HealthSyncDb, cfg: LoadedConfig, sess: requests.Session) -> str:
    if cfg.config.strava.access_token:
        return cfg.config.strava.access_token

    tok = db.get_oauth_token(STRAVA_PROVIDER)
    if not tok:
        raise RuntimeError(f"Missing Strava credentials. Run `health-sync auth strava` (config: {cfg.path}).")

    access, refresh, expires_at = tok["access_token"], tok.get("refresh_token"), _strava_expires_to_dt(tok.get("expires_at"))
    if not refresh or not expires_at or (expires_at - datetime.now(UTC)).total_seconds() > 60:
        return access

    token = request_json(
        sess,
        "POST",
        STRAVA_OAUTH_TOKEN,
        headers={"Content-Type": "application/x-www-form-urlencoded"},
        data={
            "client_id": require_str(cfg, cfg.config.strava.client_id, key="strava.client_id"),
            "client_secret": require_str(cfg, cfg.config.strava.client_secret, key="strava.client_secret"),
            "grant_type": "refresh_token",
            "refresh_token": refresh,
        },
    )
    return _store_token(db, token, refresh_fallback=refresh)


def _strava_headers(access_token: str) -> dict[str, str]:
    return {
        "Authorization": f"Bearer {access_token}",
        "User-Agent": "health-sync/0.1 (+local sqlite cache)",
        "Accept": "application/json",
    }


def _watermark_epoch(db: HealthSyncDb, *, resource: str) -> int | None:
    state = db.get_sync_state(provider=STRAVA_PROVIDER, resource=resource)
    return to_epoch_seconds(state.watermark) if state and state.watermark else None


def strava_sync(db: HealthSyncDb, cfg: LoadedConfig) -> None:
    sess = requests.Session()
    access = _strava_refresh_if_needed(db, cfg, sess)
    overlap = int(cfg.config.strava.overlap_seconds)
    page_size = max(1, min(200, int(cfg.config.strava.page_size)))

    print("Syncing Strava...")
    with db.sync_run(provider=STRAVA_PROVIDER, resource="athlete") as run:
        with db.transaction():
            athlete = request_json(sess, "GET", f"{STRAVA_BASE}/athlete", headers=_strava_headers(access))
            run.add_upsert(
                db.upsert_record(
                    provider=STRAVA_PROVIDER,
                    resource="athlete",
                    record_id=str(athlete.get("id") or "me"),
                    payload=athlete,
                    source_updated_at=utc_now_iso(),
                )
            )
            db.set_sync_state(provider=STRAVA_PROVIDER, resource="athlete", watermark=utc_now_iso())

    existing = _watermark_epoch(db, resource="activities")
    after_epoch = max(0, existing - overlap) if existing is not None else int(parse_yyyy_mm_dd(cfg.config.strava.start_date).timestamp())
    page, max_start_epoch = 1, existing
    with db.sync_run(provider=STRAVA_PROVIDER, resource="activities") as run:
        with db.transaction():
            while True:
                batch = request_json(
                    sess,
                    "GET",
                    f"{STRAVA_BASE}/athlete/activities",
                    headers=_strava_headers(access),
                    params={"after": str(after_epoch), "page": str(page), "per_page": str(page_size)},
                )
                if not isinstance(batch, list) or not batch:
                    break
                for item in batch:
                    if not isinstance(item, dict):
                        continue
                    rid = str(item.get("id") or sha256_hex(json.dumps(item, sort_keys=True, separators=(",", ":"), ensure_ascii=True)))
                    start_time = item.get("start_date")
                    start_epoch = to_epoch_seconds(start_time)
                    if start_epoch is not None and (max_start_epoch is None or start_epoch > max_start_epoch):
                        max_start_epoch = start_epoch
                    run.add_upsert(
                        db.upsert_record(
                            provider=STRAVA_PROVIDER,
                            resource="activities",
                            record_id=rid,
                            payload=item,
                            start_time=start_time,
                            source_updated_at=item.get("updated_at") or start_time,
                        )
                    )
                if len(batch) < page_size:
                    break
                page += 1
            db.set_sync_state(
                provider=STRAVA_PROVIDER,
                resource="activities",
                watermark=after_epoch if max_start_epoch is None else max_start_epoch,
            )

    print(
        "Strava sync complete "
        f"({run.inserted_count + run.updated_count + run.unchanged_count} activity records processed)."
    )

File: src/providers/strava.js
import crypto from 'node:crypto';
import {
  dtToIsoZ,
  oauthListenForCode,
  openInBrowser,
  requestJson,
  sha256Hex,
  toEpochSeconds,
  utcNowIso,
} from '../util.js';

const STRAVA_AUTH_URL = 'https://www.strava.com/oauth/authorize';
const STRAVA_TOKEN_URL = 'https://www.strava.com/oauth/token';
const STRAVA_API_BASE = 'https://www.strava.com/api/v3';

function randomState() {
  return crypto.randomBytes(16).toString('hex');
}

function parseRedirectConfig(redirectUri) {
  const parsed = new URL(redirectUri);
  return {
    host: parsed.hostname,
    port: parsed.port ? Number.parseInt(parsed.port, 10) : (parsed.protocol === 'https:' ? 443 : 80),
    path: parsed.pathname || '/callback',
    uri: parsed.toString(),
  };
}

function tokenExpiredSoon(expiresAtIso, skewSeconds = 60) {
  if (!expiresAtIso) {
    return true;
  }
  const expiryMs = Date.parse(expiresAtIso);
  if (Number.isNaN(expiryMs)) {
    return true;
  }
  return expiryMs <= (Date.now() + skewSeconds * 1000);
}

async function refreshTokenIfNeeded(db, cfgSection) {
  if (cfgSection.access_token) {
    db.setOAuthToken('strava', {
      accessToken: String(cfgSection.access_token),
      refreshToken: null,
      tokenType: 'Bearer',
      scope: cfgSection.scopes,
      expiresAt: null,
      extra: { method: 'static_access_token' },
    });
    return String(cfgSection.access_token);
  }

  const token = db.getOAuthToken('strava');
  if (!token) {
    throw new Error('Strava token not found. Run `health-sync auth strava` or set [strava].access_token.');
  }
  if (!token.refreshToken || !token.expiresAt) {
    return token.accessToken;
  }
  if (!tokenExpiredSoon(token.expiresAt)) {
    return token.accessToken;
  }

  if (!cfgSection.client_id || !cfgSection.client_secret) {
    throw new Error('Strava token expired and missing [strava].client_id/client_secret for refresh.');
  }

  const refreshed = await requestJson(STRAVA_TOKEN_URL, {
    method: 'POST',
    data: {
      client_id: cfgSection.client_id,
      client_secret: cfgSection.client_secret,
      grant_type: 'refresh_token',
      refresh_token: token.refreshToken,
    },
  });

  const expiresAt = refreshed.expires_at
    ? dtToIsoZ(new Date(Number(refreshed.expires_at) * 1000))
    : null;

  db.setOAuthToken('strava', {
    accessToken: String(refreshed.access_token),
    refreshToken: refreshed.refresh_token || token.refreshToken,
    tokenType: refreshed.token_type || 'Bearer',
    scope: refreshed.scope || token.scope,
    expiresAt,
    extra: {
      athlete: refreshed.athlete || null,
      method: 'oauth',
    },
  });

  return String(refreshed.access_token);
}

async function stravaAuth(db, config, helpers, options = {}) {
  const cfg = helpers.configFor('strava');
  if (cfg.access_token) {
    db.setOAuthToken('strava', {
      accessToken: String(cfg.access_token),
      refreshToken: null,
      tokenType: 'Bearer',
      scope: cfg.scopes,
      expiresAt: null,
      extra: { method: 'static_access_token' },
    });
    console.log('Saved static Strava access token from config.');
    return;
  }

  const clientId = helpers.requireStr('strava', 'client_id', 'Missing [strava].client_id');
  const clientSecret = helpers.requireStr('strava', 'client_secret', 'Missing [strava].client_secret');
  const redirectUri = helpers.requireStr('strava', 'redirect_uri', 'Missing [strava].redirect_uri');
  const scopes = String(cfg.scopes || 'read,activity:read_all');
  const approvalPrompt = String(cfg.approval_prompt || 'auto');

  const redirect = parseRedirectConfig(redirectUri);
  const listenHost = options.listenHost || redirect.host || '127.0.0.1';
  const listenPort = Number.isFinite(options.listenPort) && Number(options.listenPort) > 0
    ? Number(options.listenPort)
    : redirect.port;

  const state = randomState();
  const authUrl = new URL(STRAVA_AUTH_URL);
  authUrl.searchParams.set('client_id', clientId);
  authUrl.searchParams.set('response_type', 'code');
  authUrl.searchParams.set('redirect_uri', redirect.uri);
  authUrl.searchParams.set('approval_prompt', approvalPrompt);
  authUrl.searchParams.set('scope', scopes);
  authUrl.searchParams.set('state', state);

  openInBrowser(authUrl.toString());
  console.log(`Open this URL to authorize Strava: ${authUrl.toString()}`);

  const callback = await oauthListenForCode({
    listenHost,
    listenPort,
    callbackPath: redirect.path,
    timeoutSeconds: 300,
    allowManualCodeEntry: Boolean(options.allowManualCodeEntry),
    onStatus: (line) => console.log(line),
  });

  if (callback.error) {
    throw new Error(`Strava OAuth error: ${callback.error}`);
  }
  if (!callback.code) {
    throw new Error('Strava OAuth did not return an authorization code');
  }
  if (callback.state && callback.state !== state) {
    throw new Error('Strava OAuth state mismatch');
  }

  const token = await requestJson(STRAVA_TOKEN_URL, {
    method: 'POST',
    data: {
      client_id: clientId,
      client_secret: clientSecret,
      code: callback.code,
      grant_type: 'authorization_code',
    },
  });

  const expiresAt = token.expires_at
    ? dtToIsoZ(new Date(Number(token.expires_at) * 1000))
    : null;

  db.setOAuthToken('strava', {
    accessToken: String(token.access_token),
    refreshToken: token.refresh_token || null,
    tokenType: token.token_type || 'Bearer',
    scope: token.scope || scopes,
    expiresAt,
    extra: {
      athlete: token.athlete || null,
      method: 'oauth',
    },
  });

  console.log('Strava authorization succeeded.');
}

async function syncAthlete(db, token) {
  await db.syncRun('strava', 'athlete', async () => {
    await db.transaction(async () => {
      const payload = await requestJson(`${STRAVA_API_BASE}/athlete`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });

      const recordId = payload?.id ? String(payload.id) : 'me';
      db.upsertRecord({
        provider: 'strava',
        resource: 'athlete',
        recordId,
        sourceUpdatedAt: utcNowIso(),
        payload,
      });

      db.setSyncState('strava', 'athlete', { watermark: utcNowIso() });
    });
  });
}

async function syncActivities(db, token, cfg) {
  await db.syncRun('strava', 'activities', async () => {
    await db.transaction(async () => {
      const pageSize = Math.max(1, Math.min(200, Number.parseInt(String(cfg.page_size ?? 100), 10) || 100));
      const overlapSeconds = Math.max(0, Number.parseInt(String(cfg.overlap_seconds ?? 604800), 10) || 0);

      const state = db.getSyncState('strava', 'activities');
      const existingWatermarkEpoch = state?.watermark ? toEpochSeconds(state.watermark) : null;
      const startDateEpoch = toEpochSeconds(cfg.start_date || '2010-01-01') || 0;

      const afterEpoch = existingWatermarkEpoch === null
        ? startDateEpoch
        : Math.max(0, existingWatermarkEpoch - overlapSeconds);

      let page = 1;
      let maxStartEpoch = existingWatermarkEpoch === null ? afterEpoch : existingWatermarkEpoch;

      while (true) {
        const batch = await requestJson(`${STRAVA_API_BASE}/athlete/activities`, {
          headers: {
            Authorization: `Bearer ${token}`,
          },
          params: {
            after: afterEpoch,
            page,
            per_page: pageSize,
          },
        });

        const activities = Array.isArray(batch) ? batch : [];
        for (const item of activities) {
          const recordId = item?.id ? String(item.id) : sha256Hex(JSON.stringify(item));
          const startTime = item?.start_date || null;
          const startEpoch = toEpochSeconds(startTime);
          if (startEpoch !== null) {
            maxStartEpoch = Math.max(maxStartEpoch ?? startEpoch, startEpoch);
          }
          db.upsertRecord({
            provider: 'strava',
            resource: 'activities',
            recordId,
            startTime,
            endTime: null,
            sourceUpdatedAt: item?.updated_at || startTime,
            payload: item,
          });
        }

        if (activities.length < pageSize) {
          break;
        }
        page += 1;
      }

      if (maxStartEpoch !== null) {
        db.setSyncState('strava', 'activities', {
          watermark: dtToIsoZ(new Date(maxStartEpoch * 1000)),
        });
      }
    });
  });
}

async function stravaSync(db, config, helpers) {
  const cfg = helpers.configFor('strava');
  const token = await refreshTokenIfNeeded(db, cfg);
  await syncAthlete(db, token);
  await syncActivities(db, token, cfg);
}

const stravaProvider = {
  id: 'strava',
  source: 'builtin',
  description: 'Strava athlete profile and activities',
  supportsAuth: true,
  auth: stravaAuth,
  sync: stravaSync,
};

export default stravaProvider;
